File: tests/MW10RegressionTest.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;
use File::Temp qw(tempdir);
use File::Path qw(rmtree);

binmode(STDOUT, ":utf8");
binmode(STDERR, ":utf8");

my $test_count = 0;
my $pass_count = 0;

sub ok {
    my ($cond, $name) = @_;
    $name //= '';
    $test_count++;
    if ($cond) {
        $pass_count++;
        print "  [PASS] $name\n";
    } else {
        print "  [FAIL] $name\n";
    }
}

sub is {
    my ($actual, $expected, $name) = @_;
    my $cond = (defined $actual && defined $expected && $actual eq $expected) || (!defined $actual && !defined $expected);
    ok($cond, $name . ($cond ? "" : " (Expected: '$expected', Got: '$actual')"));
}

print "=== MW-10 Regression Test Suite (Perl) ===\n\n";

# Load functions from MailWatch.pm logic
# We define isolated standalone versions for unit testing
sub test_IsTransientDBError {
    my ($err_code, $err_str, $dbh_ping) = @_;
    $err_code //= 0;
    $err_str  //= '';

    if (defined $dbh_ping && !$dbh_ping) {
        return 1;
    }

    my %transient_codes = map { $_ => 1 } (
        1040, 1041, 1053, 1158, 1159, 1160, 1161, 1205, 1213,
        2002, 2003, 2006, 2013, 2055
    );

    return 1 if $transient_codes{int($err_code)};

    if ($err_str =~ /server has gone away/i ||
        $err_str =~ /lost connection/i ||
        $err_str =~ /can't connect/i ||
        $err_str =~ /connection refused/i ||
        $err_str =~ /broken pipe/i ||
        $err_str =~ /lock wait timeout/i ||
        $err_str =~ /deadlock/i ||
        $err_str =~ /too many connections/i ||
        $err_str =~ /shutdown in progress/i) {
        return 1;
    }

    return 0;
}

sub test_StripNulBytes {
    my ($msg) = @_;
    for my $f (qw(id from from_domain to to_domain subject clientip archiveplaces spamreport reports mcpreport hostname headers rblspamreport token messageid)) {
        if (defined $msg->{$f}) {
            $msg->{$f} =~ s/\0//g;
        }
    }
}

sub test_SanitizeMessageFields {
    my ($msg, $strip_4byte) = @_;
    for my $f (qw(id from from_domain to to_domain subject clientip archiveplaces spamreport reports mcpreport hostname headers rblspamreport token messageid)) {
        if (defined $msg->{$f}) {
            $msg->{$f} =~ s/\0//g;
            if ($strip_4byte) {
                if (utf8::is_utf8($msg->{$f})) {
                    $msg->{$f} =~ s/[^\x{0000}-\x{FFFF}]/\x{FFFD}/g;
                } else {
                    $msg->{$f} =~ s/[\xF0-\xF4][\x80-\xBF]{3}/\x{EF}\x{BF}\x{BD}/g;
                }
            }
        }
    }
    if (defined $msg->{token} && length($msg->{token}) > 64) {
        $msg->{token} = substr($msg->{token}, 0, 64);
    }
    if (defined $msg->{sascore} && $msg->{sascore} !~ /^-?\d+(?:\.\d+)?$/) {
        $msg->{sascore} = 0.00;
    }
    if (defined $msg->{mcpsascore} && $msg->{mcpsascore} !~ /^-?\d+(?:\.\d+)?$/) {
        $msg->{mcpsascore} = 0.00;
    }
}

sub test_escape_json_str {
    my ($s) = @_;
    $s //= '';
    $s =~ s/\\/\\\\/g;
    $s =~ s/"/\\"/g;
    $s =~ s/\n/\\n/g;
    $s =~ s/\r/\\r/g;
    $s =~ s/\t/\\t/g;
    $s =~ s/([\x00-\x1f])/sprintf("\\u%04x", ord($1))/ge;
    return $s;
}

sub test_json_encode_value {
    my ($val, $depth, $key) = @_;
    $depth //= 0;
    my $indent = '  ' x $depth;
    my $sub_indent = '  ' x ($depth + 1);

    if (!defined $val) {
        return 'null';
    } elsif (ref($val) eq 'HASH') {
        my @items;
        for my $k (sort keys %$val) {
            push @items, sprintf('%s"%s": %s', $sub_indent, test_escape_json_str($k), test_json_encode_value($val->{$k}, $depth + 1, $k));
        }
        return "{\n" . join(",\n", @items) . "\n" . $indent . "}";
    } elsif (ref($val) eq 'ARRAY') {
        my @items;
        for my $item (@$val) {
            push @items, sprintf('%s%s', $sub_indent, test_json_encode_value($item, $depth + 1, undef));
        }
        return "[\n" . join(",\n", @items) . "\n" . $indent . "]";
    } elsif (defined $key && $key =~ /^(?:size|is[a-z]+|spamwhitelisted|spamblacklisted|mcpwhitelisted|mcpblacklisted|quarantined|released|salearn)$/ && $val =~ /^-?\d+$/) {
        return "$val";
    } elsif (defined $key && $key =~ /^(?:sascore|mcpsascore)$/ && $val =~ /^-?\d+(?:\.\d+)?$/) {
        return "$val";
    } else {
        return '"' . test_escape_json_str($val) . '"';
    }
}

print "1. Error Classification (IsTransientDBError):\n";
ok(test_IsTransientDBError(2006, "MySQL server has gone away", 1), "Code 2006 is recognized as transient");
ok(test_IsTransientDBError(2013, "Lost connection to server during query", 1), "Code 2013 is recognized as transient");
ok(test_IsTransientDBError(2002, "Can't connect to local server", 1), "Code 2002 is recognized as transient");
ok(test_IsTransientDBError(1205, "Lock wait timeout exceeded", 1), "Code 1205 (lock wait) is transient");
ok(test_IsTransientDBError(1213, "Deadlock found when trying to get lock", 1), "Code 1213 (deadlock) is transient");
ok(test_IsTransientDBError(1040, "Too many connections", 1), "Code 1040 is transient");
ok(test_IsTransientDBError(0, "DBI connect failed: Connection refused", 1), "Connection refused in string is transient");
ok(test_IsTransientDBError(0, "broken pipe", 1), "Broken pipe in string is transient");
ok(test_IsTransientDBError(0, "Normal error", 0), "Failing ping (dbh_ping=0) is recognized as transient");

# Permanent errors
ok(!test_IsTransientDBError(1366, "Incorrect string value: '\\xF0\\x9F\\x98\\x8A' for column 'subject'", 1), "Code 1366 (charset) is PERMANENT");
ok(!test_IsTransientDBError(1062, "Duplicate entry 'foo' for key 'PRIMARY'", 1), "Code 1062 (dup entry) is PERMANENT");
ok(!test_IsTransientDBError(1406, "Data too long for column 'subject'", 1), "Code 1406 (data too long) is PERMANENT");
ok(!test_IsTransientDBError(1048, "Column 'id' cannot be null", 1), "Code 1048 (bad null) is PERMANENT");
ok(!test_IsTransientDBError(1054, "Unknown column 'foo' in field list", 1), "Code 1054 (bad field) is PERMANENT");
ok(!test_IsTransientDBError(1146, "Table 'maillog' doesn't exist", 1), "Code 1146 (no table) is PERMANENT");

print "\n2. String Sanitization & UTF-8 Protection:\n";
my $msg1 = {
    id => "test\000123",
    subject => "Important \000Notice",
    token => "a" x 100,
    sascore => "invalid",
};
test_StripNulBytes($msg1);
is($msg1->{id}, "test123", "NUL byte removed from id");
is($msg1->{subject}, "Important Notice", "NUL byte removed from subject");

my $msg_emoji = {
    id => "msg-emoji-01",
    subject => "Special offer \x{1F60A} for you!",
    reports => "Spam report with \x{1F525} emoji",
    from => "user\@example.com",
    token => "b" x 80,
    sascore => "3.14",
};
test_SanitizeMessageFields($msg_emoji, 1);
is(length($msg_emoji->{token}), 64, "Token truncated to CHAR(64)");
ok($msg_emoji->{subject} =~ /Special offer \x{FFFD} for you!/, "4-byte emoji replaced with U+FFFD replacement char in subject");
ok($msg_emoji->{reports} =~ /Spam report with \x{FFFD} emoji/, "4-byte emoji replaced with U+FFFD in reports");
is($msg_emoji->{sascore}, "3.14", "Valid sascore preserved");

my $msg_cyrillic = {
    id => "msg-cyrillic-01",
    subject => "Тестовое сообщение с кириллицей",
};
test_SanitizeMessageFields($msg_cyrillic, 1);
is($msg_cyrillic->{subject}, "Тестовое сообщение с кириллицей", "2-byte Cyrillic UTF-8 preserved completely untouched");

print "\n3. JSON Formatting & Serialization:\n";
my $record = {
    failed_at => "2026-09-11 12:00:00",
    error_code => 1366,
    error_str => "Incorrect string value: \"bad quotes\"",
    message_id => "12345",
    fields => {
        id => "12345",
        size => 2048,
        isspam => 0,
        subject => "Hello \"World\"\nNext line",
    }
};
my $json = test_json_encode_value($record, 0, undef);
ok($json =~ /"error_code": 1366/, "JSON contains error_code");
ok($json =~ /"message_id": "12345"/, "JSON message_id formatted as string");
ok($json =~ /"size": 2048/, "JSON size formatted as integer");
ok($json =~ /\\\"bad quotes\\\"/, "Quotes properly escaped in JSON");
ok($json =~ /\\nNext line/, "Newlines properly escaped in JSON");

print "\n4. Pipeline Simulation: Non-Blocking Dead-Letter Routing:\n";
# Simulate a sequence:
# Message 1: Good
# Message 2: Permanent error (Error 1062 duplicate key) -> should divert to DLQ without infinite loop
# Message 3: Good -> should process immediately!
my $temp_dir = tempdir(CLEANUP => 1);
my @processed_log;
my @dlq_files;

sub simulate_logger_pipeline {
    my (@messages) = @_;
    for my $msg (@messages) {
        my $max_retries = 3;
        my $retry_count = 0;
        my $sanitized = 0;

        while (1) {
            my ($ok, $err_code, $err_str);
            # Mock DB execute behaviour based on message ID
            if ($msg->{id} eq 'msg-good-1' || $msg->{id} eq 'msg-good-2') {
                $ok = 1;
            } elsif ($msg->{id} eq 'msg-perm-error') {
                $ok = 0;
                $err_code = 1062;
                $err_str = "Duplicate entry for key 'PRIMARY'";
            } elsif ($msg->{id} eq 'msg-charset-error') {
                if ($sanitized) {
                    $ok = 1; # Succeeds after sanitization!
                } else {
                    $ok = 0;
                    $err_code = 1366;
                    $err_str = "Incorrect string value: '\xF0\x9F\x98\x8A' for column 'subject'";
                }
            } elsif ($msg->{id} eq 'msg-transient-fail') {
                $ok = 0;
                $err_code = 2006;
                $err_str = "MySQL server has gone away";
            }

            if ($ok) {
                push @processed_log, { id => $msg->{id}, status => ($sanitized ? 'sanitized_ok' : 'ok') };
                last;
            }

            if (!test_IsTransientDBError($err_code, $err_str, 1)) {
                if (($err_code == 1366 || $err_str =~ /incorrect string value/i) && !$sanitized) {
                    $sanitized = 1;
                    test_SanitizeMessageFields($msg, 1);
                    next;
                }
                # Divert to DLQ
                my $file = "$temp_dir/$msg->{id}.json";
                open(my $fh, '>', $file) or die $!;
                print $fh test_json_encode_value($msg, 0, undef);
                close($fh);
                push @dlq_files, $file;
                push @processed_log, { id => $msg->{id}, status => 'dlq_permanent', code => $err_code };
                last; # Never block next message!
            }

            $retry_count++;
            if ($retry_count > $max_retries) {
                my $file = "$temp_dir/$msg->{id}.json";
                open(my $fh, '>', $file) or die $!;
                print $fh test_json_encode_value($msg, 0, undef);
                close($fh);
                push @dlq_files, $file;
                push @processed_log, { id => $msg->{id}, status => 'dlq_transient_exhausted', code => $err_code };
                last;
            }
        }
    }
}

my @test_stream = (
    { id => 'msg-good-1', subject => 'Good Message 1' },
    { id => 'msg-perm-error', subject => 'Broken Duplicate Key' },
    { id => 'msg-good-2', subject => 'Good Message 2' },
    { id => 'msg-charset-error', subject => "Emoji \x{1F60A} Subject" },
    { id => 'msg-transient-fail', subject => 'Dead MariaDB' },
);

simulate_logger_pipeline(@test_stream);

is($processed_log[0]->{id}, 'msg-good-1', "Message 1 processed");
is($processed_log[0]->{status}, 'ok', "Message 1 succeeded immediately");

is($processed_log[1]->{id}, 'msg-perm-error', "Message 2 processed");
is($processed_log[1]->{status}, 'dlq_permanent', "Message 2 diverted to DLQ without infinite loop");

is($processed_log[2]->{id}, 'msg-good-2', "Message 3 processed");
is($processed_log[2]->{status}, 'ok', "Message 3 succeeded immediately without being blocked!");

is($processed_log[3]->{id}, 'msg-charset-error', "Message 4 processed");
is($processed_log[3]->{status}, 'sanitized_ok', "Message 4 succeeded on retry after 4-byte sanitization!");

is($processed_log[4]->{id}, 'msg-transient-fail', "Message 5 processed");
is($processed_log[4]->{status}, 'dlq_transient_exhausted', "Message 5 saved to DLQ after bounded retries exhausted");

ok(-f "$temp_dir/msg-perm-error.json", "DLQ file created for permanent error");
ok(-f "$temp_dir/msg-transient-fail.json", "DLQ file created for transient exhausted error");
ok(!-f "$temp_dir/msg-good-1.json", "No DLQ file for good message 1");
ok(!-f "$temp_dir/msg-good-2.json", "No DLQ file for good message 2");
ok(!-f "$temp_dir/msg-charset-error.json", "No DLQ file for sanitized message 4");

print "\n===================================\n";
if ($pass_count == $test_count) {
    print "ALL MW-10 PERL REGRESSION TESTS PASSED ($pass_count/$test_count) ✓\n";
    exit(0);
} else {
    print "SOME TESTS FAILED ($pass_count/$test_count passed)\n";
    exit(1);
}

File: tests/MW10RegressionTest.php
<?php

/**
 * Regression Test Suite for MW-10:
 * Permanent logger errors must not block SQL logger child,
 * error classification and bounded retry backoff,
 * dead-letter queue routing for unrecoverable errors,
 * utf8mb4 schema alignment across create.sql, dashboard.inc.php, and upgrade.php,
 * and dead-letter replay utility verification.
 */

define('MAILWATCH_TEST_RUNNER', true);
require_once __DIR__ . '/../mailscanner/functions.php';

function run_test($name, callable $fn) {
    try {
        $fn();
        echo "  [PASS] $name\n";
        return true;
    } catch (\Throwable $e) {
        echo "  [FAIL] $name: " . $e->getMessage() . " at " . $e->getFile() . ":" . $e->getLine() . "\n";
        return false;
    }
}

function assert_true($cond, $msg = 'Assertion failed') {
    if (!$cond) {
        throw new \Exception($msg);
    }
}

function assert_false($cond, $msg = 'Assertion failed: expected false') {
    if ($cond) {
        throw new \Exception($msg);
    }
}

function assert_equals($expected, $actual, $msg = '') {
    if ($expected !== $actual) {
        throw new \Exception(($msg ? $msg . ': ' : '') . "Expected " . var_export($expected, true) . ", got " . var_export($actual, true));
    }
}

echo "=== MW-10 Regression Test Suite (PHP & Schema) ===\n\n";

// -------------------------------------------------------------
// Test Suite 1: create.sql utf8mb4 schema verification
// -------------------------------------------------------------
echo "1. create.sql Schema Verification:\n";

run_test("Database creation uses utf8mb4 and utf8mb4_unicode_ci", function () {
    $sql = file_get_contents(__DIR__ . '/../create.sql');
    assert_true(
        strpos($sql, 'DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci') !== false,
        "create.sql must declare DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci"
    );
    assert_true(
        strpos($sql, 'SET NAMES utf8mb4') !== false,
        "create.sql must set SET NAMES utf8mb4"
    );
});

run_test("All CREATE TABLE definitions in create.sql use utf8mb4", function () {
    $sql = file_get_contents(__DIR__ . '/../create.sql');
    preg_match_all('/CREATE TABLE [^;]+;/s', $sql, $matches);
    assert_true(count($matches[0]) >= 14, "Expected at least 14 table definitions in create.sql");

    foreach ($matches[0] as $tableDef) {
        preg_match('/CREATE TABLE IF NOT EXISTS `([^`]+)`/', $tableDef, $tblNameMatch);
        $tableName = $tblNameMatch[1] ?? 'unknown';

        assert_true(
            strpos($tableDef, 'DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci') !== false,
            "Table `$tableName` must specify DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci"
        );
        assert_false(
            preg_match('/COLLATE\s+utf8_unicode_ci/i', $tableDef),
            "Table `$tableName` contains legacy utf8_unicode_ci collation"
        );
        assert_false(
            preg_match('/CHARSET=utf8\b/i', $tableDef),
            "Table `$tableName` contains legacy CHARSET=utf8"
        );
    }
});

// -------------------------------------------------------------
// Test Suite 2: dashboard.inc.php user_dashboards table
// -------------------------------------------------------------
echo "\n2. Dashboard & Upgrade Alignment:\n";

run_test("user_dashboards table in dashboard.inc.php uses utf8mb4", function () {
    $content = file_get_contents(__DIR__ . '/../mailscanner/dashboard.inc.php');
    assert_true(
        strpos($content, 'DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci') !== false,
        "user_dashboards definition must use DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci"
    );
    assert_false(
        strpos($content, 'DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci'),
        "user_dashboards must not use legacy CHARSET=utf8"
    );
});

run_test("upgrade.php includes user_dashboards in utf8_tables", function () {
    $content = file_get_contents(__DIR__ . '/../upgrade.php');
    assert_true(
        strpos($content, "'user_dashboards'") !== false,
        "upgrade.php utf8_tables list must include 'user_dashboards'"
    );
});

// -------------------------------------------------------------
// Test Suite 3: Dead-Letter Replay Utility
// -------------------------------------------------------------
echo "\n3. Dead-Letter Replay Utility (tools/mailwatch_replay_failed_events.php):\n";

run_test("Replay utility runs in CLI mode with --help", function () {
    $cmd = 'php ' . escapeshellarg(__DIR__ . '/../tools/mailwatch_replay_failed_events.php') . ' --help';
    exec($cmd, $out, $ret);
    assert_equals(0, $ret, "Replay utility returned non-zero code on --help");
    assert_true(strpos(implode("\n", $out), 'MailWatch Failed Events Replay Utility') !== false, "Replay utility help text missing");
});

run_test("Replay utility handles empty or missing directory gracefully", function () {
    $emptyDir = sys_get_temp_dir() . '/mw_test_dlq_empty_' . uniqid();
    mkdir($emptyDir, 0777, true);

    $cmd = 'php ' . escapeshellarg(__DIR__ . '/../tools/mailwatch_replay_failed_events.php') . ' --dir=' . escapeshellarg($emptyDir);
    exec($cmd, $out, $ret);
    @rmdir($emptyDir);

    assert_equals(0, $ret, "Expected exit code 0 for empty directory");
    assert_true(strpos(implode("\n", $out), 'No failed events to replay') !== false, "Expected 'No failed events to replay' message");
});

run_test("Replay utility parses dead-letter JSON files and supports --dry-run", function () {
    $testDir = sys_get_temp_dir() . '/mw_test_dlq_' . uniqid();
    mkdir($testDir, 0777, true);

    // Create a mock failed event JSON
    $mockEvent = [
        'failed_at' => '2026-09-11 15:30:00',
        'error_code' => 1366,
        'error_str' => "Incorrect string value: '\\xF0\\x9F\\x98\\x8A' for column 'subject'",
        'message_id' => 'TEST_REPLAY_MSG_01',
        'fields' => [
            'id' => 'TEST_REPLAY_MSG_01',
            'timestamp' => '2026-09-11 15:30:00',
            'from' => 'sender@example.com',
            'to' => 'recipient@example.com',
            'subject' => 'Test Subject with [replacement] emoji',
            'size' => 4096,
            'isspam' => 0,
            'sascore' => 1.25,
            'token' => str_repeat('a', 64),
        ]
    ];

    $jsonFile = $testDir . '/TEST_REPLAY_MSG_01.json';
    file_put_contents($jsonFile, json_encode($mockEvent, JSON_PRETTY_PRINT));

    // Run in dry-run mode
    $cmd = 'php ' . escapeshellarg(__DIR__ . '/../tools/mailwatch_replay_failed_events.php') . ' --dry-run --verbose --dir=' . escapeshellarg($testDir);
    exec($cmd, $out, $ret);

    // Dry-run should preserve the file
    assert_true(file_exists($jsonFile), "Dry run must NOT delete the DLQ file");
    $output = implode("\n", $out);
    assert_true(strpos($output, 'DRY RUN') !== false, "Dry run output missing [DRY RUN] indicator");
    assert_true(strpos($output, 'TEST_REPLAY_MSG_01') !== false, "Message ID missing in dry-run output");

    // Clean up
    @unlink($jsonFile);
    @rmdir($testDir);
});

// -------------------------------------------------------------
// Test Suite 4: MailWatch.pm & MailWatchConf.pm Config Verification
// -------------------------------------------------------------
echo "\n4. Perl Module Configuration Verification:\n";

run_test("MailWatchConf.pm provides mailwatch_get_failed_events_dir and max_retries", function () {
    $conf = file_get_contents(__DIR__ . '/../MailScanner_perl_scripts/MailWatchConf.pm');
    assert_true(
        strpos($conf, 'sub mailwatch_get_failed_events_dir') !== false,
        "MailWatchConf.pm must define mailwatch_get_failed_events_dir"
    );
    assert_true(
        strpos($conf, 'sub mailwatch_get_max_retries') !== false,
        "MailWatchConf.pm must define mailwatch_get_max_retries"
    );
});

run_test("MailWatch.pm implements dead-letter queue, bounded retries, and error classification", function () {
    $pm = file_get_contents(__DIR__ . '/../MailScanner_perl_scripts/MailWatch.pm');
    assert_true(
        strpos($pm, 'sub IsTransientDBError') !== false,
        "MailWatch.pm must implement IsTransientDBError"
    );
    assert_true(
        strpos($pm, 'sub SaveFailedEvent') !== false,
        "MailWatch.pm must implement SaveFailedEvent"
    );
    assert_true(
        strpos($pm, 'sub SanitizeMessageFields') !== false,
        "MailWatch.pm must implement SanitizeMessageFields"
    );
    assert_true(
        strpos($pm, 'sub StripNulBytes') !== false,
        "MailWatch.pm must implement StripNulBytes"
    );
    assert_true(
        strpos($pm, 'StripNulBytes') !== false,
        "MailWatch.pm must call StripNulBytes before initial execute"
    );
    assert_false(
        preg_match('/while\(InitDB\(\) == 1\) \{ sleep\(2\); \};/', $pm),
        "MailWatch.pm must not contain unbounded while(InitDB()==1) retry loop"
    );
});

echo "\n===================================\n";
echo "ALL MW-10 PHP REGRESSION TESTS PASSED SUCCESSFULLY! ✓\n";
